c7n/resources/s3.py

```diff
@@ -2612,8 +2612,114 @@ def process_bucket(self, bucket):
                 raise e
 
 
+class KMSKeyResolverMixin(object):
+    """Builds a dictionary of region specific ARNs"""
+
+    arns = dict()
+
+    def resolve_keys(self, buckets):
+        if 'key' not in self.data:
+            return None
+
+        regions = {get_region(b) for b in buckets}
+        for r in regions:
+            try:
+                client = local_session(self.manager.session_factory).client('kms', region_name=r)
+                self.arns[r] = client.describe_key(
+                    KeyId=self.data.get('key')
+                ).get('KeyMetadata').get('Arn')
+            except ClientError as e:
+                self.log.error('Error validating ARNs for set-bucket-encryption: %s' % e)
+
+    def get_key(self, bucket):
+        if 'key' not in self.data:
+            return None
+        region = get_region(bucket)
+        key = self.arns.get(region)
+        if not key:
+            self.log.warning('Unable to resolve key %s for bucket %s', key, bucket.get('Name'))
+        return key
+
+
+@filters.register('bucket-encryption')
+class BucketEncryption(KMSKeyResolverMixin, Filter):
+    """Filters for S3 buckets that have bucket-encryption
+
+    :example
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-bucket-encryption-AES256
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: bucket-encryption
+                    crypto: AES256
+              - name: s3-bucket-encryption-KMS
+                resource: s3
+                region: us-east-1
+                filters
+                  - type: bucket-encryption
+                    crypto: aws:kms
+                    key: 1234abcd-12ab-34cd-56ef-1234567890ab
+
+    """
+    schema = type_schema('bucket-encryption',
+                         required=['crypto'],
+                         crypto={'type': 'string', 'enum': ['AES256', 'aws:kms']},
+                         key={'type': 'string'})
+
+    permissions = ('s3:GetBucketEncryption',)
+
+    def process(self, buckets, event=None):
+        self.resolve_keys(buckets)
+        results = []
+        with self.executor_factory(max_workers=2) as w:
+            futures = {w.submit(self.process_bucket, b): b for b in buckets}
+            for future in as_completed(futures):
+                b = futures[future]
+                try:
+                    if future.result():
```

Comment:
fwiw, the typical pattern here we use here it to check future.exception() and log and continue, before using future.result() rather than wrap in try/except block.

Reply:
Actually that pattern is incorrect.  The reason is that you will get the exception and log it but then if there is an exception future.result() will still raise the exception.  Calling `future.exection()` does not stop `future.result()` from raising.

```
for future in as_completed(futures):
  if future.exception():
    self.log.error(future.exception())
  results += filter(None, [future.result()])
return results
```

In the above example, the call to `future.result()` still causes an exception to be raised thus breaking out of the `for future` loop.

Reply:
An alternative could be to check if exception log, else append results.

Reply:
An alternative could be to check if exception log, else append results.

Reply:
hence why I said `continue` in the comment, as in next proceed to next future. the pattern is only incorrect if you use future.result() after checking thats its actually an exception and not expecting it to raise an error.

Reply:
ah right continue as in `continue`